src/coordinates/uniform_spherical.hpp

```diff
@@ -0,0 +1,158 @@
+      return r * std::sin(th);
+    } else {
+      PARTHENON_FAIL("Unknown dir");
+    }
+    return 0.0;
+  }
+
+  //----------------------------------------
+  // EdgeLength: Length of cell edges
+  //----------------------------------------
+  template <int dir>
+  KOKKOS_FORCEINLINE_FUNCTION Real EdgeLength(const int k, const int j, const int i) const {
+    static_assert(dir > 0 && dir < 4);
+    if constexpr (dir == X1DIR) {
+      // radial direction is trivial
+      return Dx<dir>();
+    } else if constexpr (dir == X2DIR) {
+      // theta direction
+      return Xf<X1DIR>(k, j, i) * Dx<dir>();
+    }
+    // phi direction
+    return Xf<X1DIR>(k, j, i) * std::sin(Xf<X2DIR>(k, j, i)) * Dx<dir>();
+  }
+  KOKKOS_FORCEINLINE_FUNCTION Real EdgeLength(const int dir, const int k, const int j, const int i) const {
+    assert(dir > 0 && dir < 4);
+    if (dir == X1DIR) return EdgeLength<X1DIR>(k, j, i);
+    else if (dir == X2DIR) return EdgeLength<X2DIR>(k, j, i);
+    return EdgeLength<X3DIR>(k, j, i);
+  }
+
+  //----------------------------------------
+  // FaceArea: Area of cell areas
+  //----------------------------------------
+  template <int dir>
+  KOKKOS_FORCEINLINE_FUNCTION Real FaceArea(const int k, const int j, const int i) const {
+    static_assert(dir > 0 && dir < 4);
+    if constexpr (dir == X1DIR) {
+      Real dOmega = Dx<X3DIR>() * (std::cos(Xf<X2DIR>(k, j, i)) - std::cos(Xf<X2DIR>(k, j+1, i)));
+      Real r = Xf<X1DIR>(k, j, i);
+      return r * r * dOmega;
+    } else if constexpr (dir == X2DIR) {
+      Real r0 = Xf<X1DIR>(k, j, i);
+      Real r1 = Xf<X1DIR>(k, j, i+1);
+      return (r1*r1*r1 - r0*r0*r0) * std::sin(Xf<X2DIR>(k, j, i)) * Dx<X3DIR>() / 3.0;
+    }
+    Real r0 = Xf<X1DIR>(k, j, i);
+    Real r1 = Xf<X1DIR>(k, j, i+1);
+    Real dcth = std::cos(Xf<X2DIR>(k, j, i)) - std::cos(Xf<X2DIR>(k, j+1, i));
+    return (r1*r1*r1 - r0*r0*r0) * dcth / 3.0;
+  }
+
+  //----------------------------------------
+  // CellVolume
+  //----------------------------------------
+  KOKKOS_FORCEINLINE_FUNCTION Real CellVolume(const int k, const int j, const int i) const {
+    return FaceArea<X3DIR>(k, j, i) * Dx<X3DIR>();
+  }
+
+  KOKKOS_FORCEINLINE_FUNCTION
+  Real Volume(CellLevel cl, TopologicalElement el, const int k, const int j, const int i) {
+    using TE = TopologicalElement;
+    if (cl == CellLevel::same) {
+      if (el == TE::CC) return CellVolume(k, j, i);
+      else if (el == TE::F1) return FaceArea<X1DIR>(k, j, i);
+      else if (el == TE::F2) return FaceArea<X2DIR>(k, j, i);
+      else if (el == TE::F3) return FaceArea<X3DIR>(k, j, i);
+      else if (el == TE::E1) return EdgeLength<X1DIR>(k, j, i);
+      else if (el == TE::E2) return EdgeLength<X2DIR>(k, j, i);
+      else if (el == TE::E3) return EdgeLength<X3DIR>(k, j, i);
+      else if (el == TE::NN) return 1.0;
+    } else {
+      PARTHENON_FAIL("Have not yet implemented fine fields for UniformSpherical coordinates.");
```

Comment:
We could just have a coordinates object own another coordinates object that has twice the number of zones and then just call that for fine fields.